include/pybind11/pybind11.h

```diff
@@ -908,11 +908,24 @@ inline void call_operator_delete(void *p) { ::operator delete(p); }
 
 NAMESPACE_END(detail)
 
+/// Given a pointer to a member function, cast it to its `Derived` version.
+/// Forward everything else unchanged.
+template <typename /*Derived*/, typename F>
+auto method_adaptor(F &&f) -> decltype(std::forward<F>(f)) { return std::forward<F>(f); }
+
+template <typename Derived, typename Return, typename Class, typename... Args,
+          typename Adapted = Return (Derived::*)(Args...)>
+Adapted method_adaptor(Return (Class::*pmf)(Args...)) { return pmf; }
```

Comment:
`Adapted` is only there to avoid the horror of the regular function pointer return type syntax. But the trailing return type solution looks much nicer!